### PR TITLE
doc/rados: Edit troubleshooting-osd (2 of x)

### PR DESCRIPTION
**Pull request by zdover23**
Edit doc/rados/troubleshooting/troubleshooting.rst (2 of x).

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "pacific"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [x] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 1]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
 An OSD Failed
 -------------
 
-When a ``ceph-osd`` process dies, surviving ``ceph-osd`` daemons will report
-to the mons that it appears down, which will in turn surface the new status
-via the ``ceph health`` command::
+When an OSD has failed, this means that a ``ceph-osd`` process has died and
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:270]: s/has failed/fails/
s/has died/is unresponsive or has died/

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 2]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
 
-	ceph health
-	HEALTH_WARN 1/3 in osds are down
+.. prompt:: bash
+
+   ceph health
+
+::
+
+   HEALTH_WARN 1/3 in osds are down
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:282]: It pains me that this message is lowercase, but the source verifies.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 3]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+
+This health alert is raised whenever there are one or more OSDs marked ``in``
+and ``down``. To see which OSDs are ``down``, add the ``detail`` command as in
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:285]: add ``detail`` to the command

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 4]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
-If there is a drive
-failure or other fault preventing ``ceph-osd`` from functioning or
-restarting, an error message should be present in its log file under
-``/var/log/ceph``.
+If the ``ceph-osd`` daemon stopped because of a heartbeat failure or a
+``suicide timeout`` error, then the underlying drive or filesystem might be
+unresponsive. Check ``dmesg`` output and `syslog`  output for drive errors or
+kernel errors. It might be necessary to specify certain flags (for example,
+``dmesg -T`` to see human-readable timestamps) in order to avoid mistaking old
+errors for new errors.
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:312]: Maybe add a note that if an entire host's OSDs are `down`, there might either be a hardware issue with the host or a network issue.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 5]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
 
-Ceph prevents you from writing to a full OSD so that you don't lose data.
-In an operational cluster, you should receive a warning when your cluster's OSDs
-and pools approach the full ratio. The ``mon_osd_full_ratio`` defaults to
-``0.95``, or 95% of capacity before it stops clients from writing data.
-The ``mon_osd_backfillfull_ratio`` defaults to ``0.90``, or 90 % of
-capacity above which backfills will not start. The
-OSD nearfull ratio defaults to ``0.85``, or 85% of capacity
-when it generates a health warning.
+If an OSD is full, Ceph prevents data loss by ensuring that nothing is written
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:329]: that no new data is written

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 6]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+to the OSD. In an properly running cluster, health checks are raised when the
+cluster's OSDs and pools approach certain "fullness" ratios. The
+``mon_osd_full_ratio`` threshold defaults to ``0.95`` (or 95% of capacity):
+this is the point above which clients are prevented from writing data. The
+``mon_osd_backfillfull_ratio`` threshold defaults to ``0.90`` (or 90% of
+capacity): this is the point above which backfills will not start. The
+``mon_osd_nearfull_ratio`` threshold defaults to ``0.85`` (or 85% of capacity):
+this is the point at which it raises the ``OSD_NEARFULL`` health check.
+
+Different OSDs within a cluster will vary in how much data is allocated to them
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:339]: s/Different//

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 7]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+When examining the output of the ``ceph df`` command, pay special attention to
+the **most full** OSDs, as opposed to the percentage of raw space used. If a
+single outlier OSD becomes full, all writes to this OSD's pool might fail as a
+result. When ``ceph df`` reports the space available to a pool, it considers
+the ratio settings relative to the *most full* OSD that is part of the pool. To
+flatten the distribution, two approaches are available: (1) Using the
+``reweight-by-utilization`` command to progressively move data from excessively
+full OSDs or move data to insufficiently full OSDs, and (2) in later revisions
+of Luminous and subsequent releases, exploiting the ``ceph-mgr`` ``balancer``
+module to perform the same task automatically.
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:363]: Maybe it's time to leave r-b-u in the past.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 8]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+   ceph osd set-backfillfull-ratio <float[0.0-1.0]>
+
+Sometimes full cluster issues arise because an OSD has failed. This can happen
+either because of a test or because the cluster is small, very full, or
+unbalanced. When an OSD or node holds an excessive percentage of the cluster's
+data, component failures or natural growth can result in the ``nearfull`` and
+``full`` ratios being exceeded.  Anyone testing how Ceph reacts to OSD failures
+on a small cluster is advised to leave ample free disk space and to consider
+temporarily lowering the OSD ``full ratio``, OSD ``backfillfull ratio``, and
+OSD ``nearfull ratio``.
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:381]: When testing Ceph resilience to OSD failures on a small cluster, it is

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 9]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+    HEALTH_ERR 1 full osd(s); 1 backfillfull osd(s); 1 nearfull osd(s)
+    osd.3 is full at 97%
+    osd.4 is backfill full at 91%
+    osd.2 is near full at 87%
 
-The best way to deal with a full cluster is to add capacity via new OSDs, enabling
-the cluster to redistribute data to newly available storage.
+To address full cluster issues, it is recommended to add capacity by adding
+OSDs.  Adding new OSDs allows the cluster to redistribute data to newly
+available storage.
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:409]: Perhaps add a note to search for `rados bench` orphans that are wasting space.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 10]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
-   **DO NOT** delete the same placement group directory on another full OSD, or
-   **YOU WILL LOSE DATA**. You **MUST** maintain at least one copy of your data on
-   at least one OSD.  This is a rare and extreme intervention, and is not to be
-   undertaken lightly.
+.. important:: If you choose to delete a placement group directory on a full
+   OSD, **DO NOT** delete the same placement group directory on another full
+   OSD. **OTHERWISE YOU WILL LOSE DATA**. You **MUST** maintain at least one
+   copy of your data on at least one OSD. Deleting placement group directories
+   is a rare and extreme intervention. It is not to be undertaken lightly.
+
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:420]: Maybe it's time to drop this Filestore section.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 11]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
-See `Monitor Config Reference`_ for additional details.
 
 OSDs are Slow/Unresponsive
 ==========================
 
-A common issue involves slow or unresponsive OSDs. Ensure that you
-have eliminated other troubleshooting possibilities before delving into OSD
-performance issues. For example, ensure that your network(s) is working properly
-and your OSDs are running. Check to see if OSDs are throttling recovery traffic.
+OSDs are sometimes slow or unresponsive. Anyone troubleshooting this common
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:427]: s/Anyone/When/

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 12]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+problem is advised to eliminate other possibilities before investigating OSD
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:428]: it is advised
or
we advise eliminating

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 13]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+performance issues. For example, be sure to confirm that your network(s) are
+working properly, to verify that your OSDs are running, and to check whether
+OSDs are throttling recovery traffic.
+
+.. tip:: In pre-Luminous releases of Ceph, ``up`` and ``in`` OSDs were
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:433]: Drop?

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 14]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
-   OSDs are not available or are otherwise slow.
 
 Networking Issues
 -----------------
 
-Ceph is a distributed storage system, so it relies upon networks for OSD peering
-and replication, recovery from faults, and periodic heartbeats. Networking
-issues can cause OSD latency and flapping OSDs. See `Flapping OSDs`_ for
-details.
+Because Ceph is a distributed storage system, it relies upon networks for OSD
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:442]: As a distributed system, Ceph relies

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 15]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
 
-	netstat -s
+.. prompt:: bash
+
+   netstat -s
 
 Drive Configuration
 -------------------
 
-A SAS or SATA storage drive should only house one OSD; NVMe drives readily
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:415]: With Quincy and Reef, @markhpc has shown that the performance advantages of multiple OSDs per NVMe device have mostly been obviated, so perhaps we should update this to only suggest one OSD per drive, period.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 16]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+journals / metadata, operating systems, Ceph monitors, ``syslog`` logs, other
+OSDs, and non-Ceph processes.
 
-Ceph acknowledges writes *after* journaling, so fast SSDs are an
-attractive option to accelerate the response time--particularly when
-using the ``XFS`` or ``ext4`` file systems for legacy Filestore OSDs.
-By contrast, the ``Btrfs``
-file system can write and journal simultaneously.  (Note, however, that
-we recommend against using ``Btrfs`` for production deployments.)
+Because Ceph acknowledges writes *after* journaling, fast SSDs are an
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:471]: Suggest taking out lines 471-480.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 17]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
 -----------------------------
 
-Check your drives for bad blocks, fragmentation, and other errors that can cause
-performance to drop substantially.  Invaluable tools include ``dmesg``, ``syslog``
-logs, and ``smartctl`` (from the ``smartmontools`` package).
+Check your drives for bad blocks, fragmentation, and other errors that can
+cause significantly degraded performance. Tools that are useful in checking for
+drive errors include ``dmesg``, ``syslog`` logs, and ``smartctl`` (found in the
+``smartmontools`` package).
+
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:494]: Suggest ``smartmontools`` 7.0 or later for NVMe stat passthrough and JSON output.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 18]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
-same drive as an OSD. Additionally, if you run monitors on the same host as
-OSDs, you may incur performance issues related to:
-
-- Running an older kernel (pre-3.0)
-- Running a kernel with no ``syncfs(2)`` syscall.
+Although monitors are relatively lightweight processes, performance issues can
+result when monitors are run on the same host machine as an OSD. Monitors issue
+many ``fsync()`` calls and this can interfere with other workloads. The danger
+of performance issues is especially acute when the monitors are co-resident
+with an OSD. In addition, if the monitors are running an older kernel (pre-3.0)
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:503]: I think this co-residency implies on the same storage drive as an OSD, which should be specified.

File: doc/rados/troubleshooting/troubleshooting-osd.rst [hunk 19]
```diff
@@ -267,206 +267,270 @@ If the cluster has started but an OSD isn't starting, check the following:
...
+same hardware as OSDs. For this reason, making such processes co-resident with
+OSDs is not generally recommended. Instead, the recommended practice is to
+optimize certain hosts for use with Ceph and use other hosts for other
+processes. This practice of separating Ceph operations from other applications
+might help improve performance and might also streamline troubleshooting and
+maintenance.
+
+Running co-resident processes on the same hardware is sometimes called
+"convergence". When using Ceph, avoid convergence.
+
```
Comment (anthonyeleven) [inline doc/rados/troubleshooting/troubleshooting-osd.rst:523]: This kind of flies in the face of Proxmox.  Perhaps we could recast a bit as a caveat.

**Review by anthonyeleven — APPROVED** [review]

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 1 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:270]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 3 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:285]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 4 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:312]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 5 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:329]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 6 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:339]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 7 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:363]: I will remove this in a near-future PR separate from this PR.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 8 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:381]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 9 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:409]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 10 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:420]: I will do this in a near-future pull request.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 11 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:427]: Accepted with slight modifications.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 14 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:442]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 15 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:415]: I will update this in a near-future pull request.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 17 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:494]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 18 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:503]: Accepted.

File: doc/rados/troubleshooting/troubleshooting-osd.rst (hunk 19 above)
Comment (zdover23) [reply doc/rados/troubleshooting/troubleshooting-osd.rst:523]: Accepted.

**Comment by zdover23** [comment]
https://github.com/ceph/ceph/pull/54000 - Reef backport
https://github.com/ceph/ceph/pull/54001- Quincy backport

**Pull request merged**